components/file_system/src/rate_limiter.rs

```diff
@@ -1,148 +1,570 @@
+        do_sleep!(wait, $mode);
+        tls_collect_rate_limiter_request_wait($priority.as_str(), wait);
+        amount
+    }};
+}
+
+impl PriorityBasedIORateLimiter {
+    fn new() -> Self {
+        PriorityBasedIORateLimiter {
+            bytes_through: Default::default(),
+            bytes_per_epoch: Default::default(),
+            protected: Mutex::new(PriorityBasedIORateLimiterProtected::new()),
+        }
+    }
+
+    /// Dynamically changes the total IO flow threshold.
     #[allow(dead_code)]
-    pub fn reset(&self) {
-        for i in self.read.iter() {
-            i.store(0, Ordering::Relaxed);
+    fn set_bytes_per_sec(&self, bytes_per_sec: usize) {
+        let now = (bytes_per_sec as f64 * DEFAULT_REFILL_PERIOD.as_secs_f64()) as usize;
+        let before = self.bytes_per_epoch[IOPriority::High as usize].swap(now, Ordering::Relaxed);
+        if before == 0 || now == 0 {
+            // toggle on/off rate limit.
+            // we hold this lock so a concurrent refill can't negate our effort.
+            let _locked = self.protected.lock();
+            for p in &[IOPriority::Medium, IOPriority::Low] {
+                let pi = *p as usize;
+                self.bytes_per_epoch[pi].store(now, Ordering::Relaxed);
+            }
         }
-        for i in self.write.iter() {
-            i.store(0, Ordering::Relaxed);
+    }
+
+    fn request(&self, priority: IOPriority, amount: usize) -> usize {
+        request_imp!(self, priority, amount, sync)
+    }
+
+    async fn async_request(&self, priority: IOPriority, amount: usize) -> usize {
+        request_imp!(self, priority, amount, async)
+    }
+
+    #[cfg(test)]
+    fn request_with_skewed_clock(&self, priority: IOPriority, amount: usize) -> usize {
+        request_imp!(self, priority, amount, skewed_sync)
+    }
+
+    /// Update and refill IO budgets for next epoch based on IO priority.
+    /// Here we provide best-effort priority control:
+    /// 1) IO budget is assigned to lower priority to make sure higher priority can always consume
+    ///    the same IO amount as the last few epochs within global threshold.
+    /// 2) Higher priority can temporarily use lower priority's IO budgets. When this happens,
+    ///    total IO flow will exceed global threshold.
+    /// 3) Highest priority IO alone must not exceed global threshold.
+    fn refill(&self, locked: &mut MutexGuard<PriorityBasedIORateLimiterProtected>, now: Instant) {
+        const UPDATE_BUDGETS_EVERY_N_EPOCHS: usize = 5;
+        // Epochs can be skipped for refill, which happens fairly rare under production workload.
+        // We will ignore these epochs when making decision.
+        // assuming clock can be skewed for no more than half an epoch
+        let skipped_epochs = ((now - locked.next_refill_time).as_secs_f64()
+            / DEFAULT_REFILL_PERIOD.as_secs_f64()
+            + 0.5) as usize;
```

Comment:
Right now the unit tests could pass, but I'm still not very confident in capturing all the corner cases. Generally I'm leaning towards using a separate daemon refill thread so to not worry about `skipped_epochs` or other unrevealed racy cases.

Reply:
If it simplifies the implementation, let's do that. but how about use a separate thread, instead of the bg_worker pool?

Reply:
I think the unreproducible bugs in CI has been fixed now. Maybe a separate refiller isn't needed after all.